### PR TITLE
GH-99380: Update to Sphinx 7

### PR DESCRIPTION
**Pull request by AA-Turner**
Preview: https://cpython-previews--99381.org.readthedocs.build/en/99381/

* Issue: gh-99380

### CODE REVIEW

**Comment by AA-Turner** [comment]
Rendered: https://github.com/python/cpython/suites/9258549124/artifacts/433345867

**Comment by hugovk** [comment]
With the PR, there some extra lines in the footer, extending from the sidebar collapser thing (Chrome 106, macOS).

PR on left, `main` on right:

image

**Comment by AA-Turner** [comment]
Oh, it looks like two sidebar buttons are being added. In which case we revert to `python-docs-theme` for the JS/CSS changes to be merged...

A

**Comment by AA-Turner** [comment]
I've temporarily updated the ``python-docs-theme`` version to point at https://github.com/python/python-docs-theme/pull/127, to demonstrate the fix in that PR.

A

**Comment by CAM-Gerlach** [comment]
Especially before we backport this change to 3.11, we should add a job to build the docs on the minimum Sphinx version specified in the `conf.py`, so we can continue to verify everything will build (without warnings) on the minimum version we support redistributors using, as we have for the past. This will allow us to be more flexible with keeping all supported docs versions in sync and up to date with the latest Sphinx changes and improvements.

We also have to consider if we want to retain the now opt-out option for listing functions, classes, etc. in the toctree enabled, since this PR effectively flips that switch. Maybe it should be disabled for now and enabled in a separate PR after appropriate discussion (on Discourse, etc)?

**Comment by AA-Turner** [comment]
We've spoken about a minimum build for a while -- whilst I don't think it should block this work, I'll open a new PR to that effect (do you think we need a new issue to discuss that?).

I'm biased, as I added the feature, but I think it is a real usability win to have the direct links, and so long as the theme responds reasonably (eg on long names), which I'll test, I think we should keep the defaults as a net benefit (especially in reference-type documentation).

A

**Comment by AA-Turner** [comment]
Cross-linking #103904 for the oldest-supported testing.

A

**Comment by AA-Turner** [comment]
To resolve: ToC directives, e.g. https://deploy-preview-99381--python-cpython-preview.netlify.app/library/index.html

image

File: Doc/tools/extensions/pyspecific.py [hunk 1]
```diff
@@ -21,7 +21,6 @@
 from docutils.utils import new_document
 from sphinx import addnodes
 from sphinx.builders import Builder
-from sphinx.domains.python import PyFunction, PyMethod
 from sphinx.errors import NoUri
```
Comment (hugovk) [inline Doc/tools/extensions/pyspecific.py]: These are used:
```pytb
  File "/home/runner/work/cpython/cpython/Doc/tools/extensions/pyspecific.py", line 342, in <module>
    class PyDecoratorFunction(PyDecoratorMixin, PyFunction):
                                                ^^^^^^^^^^
NameError: name 'PyFunction' is not defined
```


```suggestion
from sphinx.domains.python import PyFunction, PyMethod
from sphinx.errors import NoUri
```

Comment (AA-Turner) [reply Doc/tools/extensions/pyspecific.py]: Oops, merge error--will resolve when I'm back at my computer!

Thanks for catching it :)

A

File: Doc/requirements.txt [hunk 2]
```diff
@@ -6,8 +6,7 @@
 # Sphinx version is pinned so that new versions that introduce new warnings
 # won't suddenly cause build failures. Updating the version is fine as long
 # as no warnings are raised by doing so.
-# PR #104777: Sphinx 6.2 no longer uses imghdr, removed in Python 3.13.
-sphinx==6.2.1
+sphinx~=7.1.0
```
Comment (hugovk) [inline Doc/requirements.txt]: We could do with Sphinx 7 for this fix: https://github.com/python/cpython/pull/110885#discussion_r1359845086

Let's bump and re-review, now the footer lines are fixed, and we're testing CI with the oldest supported Sphinx.

```suggestion
sphinx~=7.2.0
```

**Pull request merged**